app/controllers/users/passwords_controller.rb

```diff
@@ -0,0 +1,42 @@
+class Users::PasswordsController < ApplicationController
+  before_action :authenticated?, only: [:edit, :update]
+
+  def edit
+    @user = current_user
+  end
+
+  def reset
+    ConfirmationSender.send_confirmation_to(current_user)
+    redirect_to new_confirmation_path
+  end
+
+  def update
+    require "pry"; binding.pry
+    if passwords_not_empty? && passwords_equal?
+      current_user.update(password_params)
+      redirect_to users_dashboard_path(current_user.username), success: "Password Updated"
+      session[:authenticated] = false
+    else
+      redirect_to users_password_edit_path(current_user.username), warning: "Error, please try again."
+    end
+  end
+
+  private
+
+  def authenticated?
+    render :file => "#{Rails.root}/public/404.html", :status => 404 unless session[:authenticated]
+  end
+
+  def password_params
+    params.require(:user).permit(:password, :password_confirmation)
+  end
+
+  def passwords_not_empty?
```

Comment:
i'd suggest moving this logic out of the controller. you could create a PORO that checks password lengths and equality or find a way to put it in the User model. the controller doesn't need to know about all of this logic